tomcat/src/main/java/com/linecorp/armeria/server/tomcat/TomcatService.java

```diff
@@ -254,125 +240,21 @@ static String toString(org.apache.catalina.Server server) {
-        this(hostname, connectorFactory, unused -> { /* unused */ });
-    }
-
-    private TomcatService(@Nullable String hostname,
-                          Function<String, Connector> connectorFactory, Consumer<Connector> postStopTask) {
-
-        this.hostname = hostname;
-        this.connectorFactory = connectorFactory;
-        this.postStopTask = postStopTask;
-        configurator = new Configurator();
-    }
-
-    @Override
-    public void serviceAdded(ServiceConfig cfg) throws Exception {
-        if (hostname == null) {
-            hostname = cfg.server().defaultHostname();
-        }
-
-        if (armeriaServer != null) {
-            if (armeriaServer != cfg.server()) {
-                throw new IllegalStateException("cannot be added to more than one server");
-            } else {
-                return;
-            }
-        }
-
-        armeriaServer = cfg.server();
-        armeriaServer.addListener(configurator);
-    }
-
     /**
      * Returns Tomcat {@link Connector}.
      */
-    public Connector connector() {
-        final Connector connector = this.connector;
-        if (connector == null) {
-            throw new IllegalStateException("not started yet");
-        }
-
-        return connector;
-    }
-
-    void start() throws Exception {
-        assert hostname != null;
-        started = false;
-        connector = connectorFactory.apply(hostname);
-        final Service service = connector.getService();
-        if (service == null) {
-            return;
-        }
-
-        final Engine engine = TomcatUtil.engine(service, hostname);
-        final String engineName = engine.getName();
-        if (engineName == null) {
-            return;
-        }
-
-        if (activeEngines.contains(engineName)) {
-            throw new TomcatServiceException("duplicate engine name: " + engineName);
-        }
-
-        server = service.getServer();
-
-        if (!TOMCAT_START_STATES.contains(server.getState())) {
-            logger.info("Starting an embedded Tomcat: {}", toString(server));
-            server.start();
-            started = true;
-        }
-
-        activeEngines.add(engineName);
-        this.engineName = engineName;
-    }
-
-    void stop() throws Exception {
-        final org.apache.catalina.Server server = this.server;
-        final Connector connector = this.connector;
-        this.server = null;
-        this.connector = null;
-
-        if (engineName != null) {
-            activeEngines.remove(engineName);
-            engineName = null;
-        }
-
-        if (server == null || !started) {
-            return;
-        }
-
-        try {
-            logger.info("Stopping an embedded Tomcat: {}", toString(server));
-            server.stop();
-        } catch (Exception e) {
-            logger.warn("Failed to stop an embedded Tomcat: {}", toString(server), e);
-        }
+    @Nullable
+    public abstract Connector connector();
```

Comment:
The previous behavior of this method was to throw `IllegalStateException` if `connector` is `null`. Would it be possible to keep the old behavior and remove `@Nullable`?

Reply:
My impression from https://github.com/line/armeria/pull/1249/files#diff-d20f3c660ecfc12f2baf4a22607a08a2L374
was that we wanted to detect this situation and `throw HttpStatusException.of(HttpStatus.SERVICE_UNAVAILABLE)`?

Reply:
https://github.com/line/armeria/pull/1249/files#diff-d20f3c660ecfc12f2baf4a22607a08a2R259

Reply:
I see. How about then returning an `Optional<Connector>`? It's a breaking change anyway.